src/source/src/dml_manager.rs

```diff
@@ -32,35 +34,63 @@ pub type DmlManagerRef = Arc<DmlManager>;
 /// channel instead of offering a `write_chunk` interface).
 #[derive(Default, Debug)]
 pub struct DmlManager {
-    batch_dmls: Mutex<HashMap<TableId, TableSourceRef>>,
+    batch_dmls: RwLock<HashMap<TableId, Weak<TableSource>>>,
 }
 
 impl DmlManager {
     pub fn new() -> Self {
         Self {
-            batch_dmls: Mutex::new(HashMap::new()),
+            batch_dmls: RwLock::new(HashMap::new()),
         }
     }
 
     pub fn register_reader(
         &self,
-        table_id: &TableId,
+        table_id: TableId,
         column_descs: &[ColumnDesc],
-    ) -> TableSourceRef {
-        let mut batch_dmls = self.batch_dmls.lock();
-        if !batch_dmls.contains_key(table_id) {
-            let batch_dml = Arc::new(TableSource::new(column_descs.to_vec()));
-            batch_dmls.insert(*table_id, batch_dml);
+    ) -> Result<TableSourceRef> {
+        let mut batch_dmls = self.batch_dmls.write();
+        match batch_dmls.entry(table_id) {
+            Entry::Occupied(o) => o.get().upgrade().ok_or_else(|| {
+                InternalError(format!(
+                    "fail to register reader for table with id {:?}",
+                    table_id.table_id
+                ))
+                .into()
+            }),
+            Entry::Vacant(v) => {
+                let reader = Arc::new(TableSource::new(column_descs.to_vec()));
+                v.insert(Arc::downgrade(&reader));
+                Ok(reader)
+            }
         }
-        batch_dmls.get(table_id).unwrap().clone()
     }
 
     pub fn write_chunk(
         &self,
         table_id: &TableId,
         chunk: StreamChunk,
     ) -> Result<oneshot::Receiver<usize>> {
-        let batch_dmls = self.batch_dmls.lock();
-        batch_dmls.get(table_id).unwrap().write_chunk(chunk)
+        let batch_dmls = self.batch_dmls.read();
+        let writer = batch_dmls
+            .get(table_id)
+            .ok_or_else(|| {
+                InternalError(format!(
+                    "fail to write into table with id {:?}",
+                    table_id.table_id
+                ))
+            })?
+            .upgrade()
+            .ok_or_else(|| {
+                InternalError(format!(
+                    "fail to write into table with id {:?}",
```

Comment:
How about making this message more specific? Like "all readers have been dropped".